h2o-algos/src/main/java/hex/hglm/HGLM.java

```diff
@@ -0,0 +1,365 @@
+    init(false);
+  }
+
+  public HGLM(HGLMModel.HGLMParameters parms, Key<HGLMModel> key) {
+    super(parms, key);
+    init(false);
+  }
+
+  @Override
+  protected ModelBuilder<HGLMModel, HGLMModel.HGLMParameters, HGLMModel.HGLMModelOutput>.Driver trainModelImpl() {
+    return new HGLMDriver();
+  }
+  
+  static class ScoringHistory {
+    private ArrayList<Integer> _scoringIters = new ArrayList<>();
+    private ArrayList<Long> _scoringTimes = new ArrayList<>();
+    private ArrayList<Double> _logLikelihood = new ArrayList<>();
+    private ArrayList<Double> _tauEVar = new ArrayList<>();
+
+    public ArrayList<Integer> getScoringIters() { return _scoringIters;}
+    public ArrayList<Long> getScoringTimes() { return _scoringTimes;}
+    public ArrayList<Double> getLikelihoods() { return _logLikelihood;}
+    public ArrayList<Double> getTauEVar() { return _tauEVar; }
+    
+    public void addIterationScore(int iter, double loglikelihood, double tauEVar) {
+      _scoringIters.add(iter);
+      _scoringTimes.add(System.currentTimeMillis());
+      _logLikelihood.add(loglikelihood);
+      _tauEVar.add(tauEVar);
+    }
+    
+    public TwoDimTable to2dTable() {
+      String[] cnames = new String[]{"timestamp", "number_of_iterations", "loglikelihood", "noise_variance"};
+      String[] ctypes = new String[]{"string", "int", "double", "double"};
+      String[] cformats = new String[]{"%s", "%d", "%.5f", "%.5f"};
+      int tableSize = _scoringIters.size();
+      TwoDimTable res = new TwoDimTable("Scoring History", "", 
+              new String[tableSize], cnames, ctypes, cformats, "");
+      int col = 0;
+      for (int i=0; i<tableSize; i++) {
+        res.set(i, col++, DATE_TIME_FORMATTER.print(_scoringTimes.get(i)));
+        res.set(i, col++, _scoringIters.get(i));
+        res.set(i, col++, _logLikelihood.get(i));
+        res.set(i, col, _tauEVar.get(i));
+        col = 0;
+      }
+      return res;
+    }
+  }
+
+  @Override
+  public void init(boolean expensive) {
+    if (_parms._nfolds > 0 || _parms._fold_column != null)
+      error("nfolds or _fold_coumn", " cross validation is not supported in HGLM right now.");
+
+    if (null != _parms._family && !gaussian.equals(_parms._family))
+      error("family", " only Gaussian families are supported now");
+
+    if (null != _parms._method && !EM.equals(_parms._method))
+      error("method", " only EM (expectation maximization) is supported for now.");
+
+    if (null != _parms._missing_values_handling &&
+            PlugValues == _parms._missing_values_handling && _parms._plug_values == null)
+      error("PlugValues", " if specified, must provide a frame with plug values in plug_values.");
+
+    if (_parms._tau_u_var_init < 0)
+      error("tau_u_var_init", "if set, must > 0.0.");
+
+    if (_parms._tau_e_var_init < 0)
+      error("tau_e_var_init", "if set, must > 0.0.");
+
+    if (_parms._seed == 0)
+      error("seed", "cannot be set to any number except zero.");
```

Comment:
Here is an example of random number generator:
![image](https://github.com/user-attachments/assets/56751351-b0ad-4f91-806e-f0c634c47e6a)

The seed is the sequence that go into the squares (delay gates).  If that seed is all 0, the output will be 0 as well.  A history of my communications background.

Reply:
Thanks for the explanation. It makes sense to me. I am surprised I have never seen this seed check in other algorithms. :thinking:  But maybe it is because I was out of work too long. ;)